Initiate targets pipeline

lekoenig opened the pull request:
Addresses #5.

After reading through @padilla410's reviews of wqp pull patterns in existing USGS-R repos, I created a `_targets.R` file where I defined the spatial and temporal extents of the WQP pull and specified parameter families of interest in our pipeline (all of these values are placeholders and can be changed moving forward). 

I also started a `1_fetch.R` file that reads in a yml file containing the parameter families and WQP `characteristicNames`, but I planned to hold off on starting any data downloads until we sketch out our pipeline phases a bit further. Currently, the most downstream output is a list containing the `characteristicNames` for the parameter families of interest:
```
> tar_load(p1_charNames)
> p1_charNames
$nitrate
 [1] "Nitrate"                                       "Nitrate + Nitrite"                             "Nitrate as N"                                 
 [4] "Nitrate-N"                                     "Nitrate-Nitrite"                               "Nitrate-Nitrogen"                             
 [7] "Nitrate-nitrogen"                              "Nitrite"                                       "Nitrite as N"                                 
[10] "Inorganic nitrogen (nitrate and nitrite)"      "Inorganic nitrogen (nitrate and nitrite) as N" "Nitrite-N"                                    

$conductivity
[1] "Specific conductivity"                           "Specific conductance"                           
[3] "Specific Conductance, Calculated/Measured Ratio" "Conductivity"                                   

> 
```

This PR is meant to be a jumping off point for further discussion and I'm open to any changes or suggestions. One question I had, for example, was whether we wanted to just use specific `characteristicNames` (e.g. "Specific conductance"), or to pull parameter families (to at least show how many unique `characteristicNames` might be tied to a single parameter of interest. 

Thanks!

padilla410 commented on .gitignore:
```diff
@@ -1,6 +1,17 @@
 .RData
-.Ruserdata
 
-*.Rproj
```
@lekoenig I see that you have removed the `*.Rproj` file from the `.gitignore`. However, when I pulled this branch locally I did not see `ds-pipelines-targets-example-wqp.Rproj`. 

Can you add another commit that adds this file back in?

padilla410 reviewed (changes requested):
Looks good so far. Please add the `.Rproj` file in another commit. Then I will run your pipeline to verify that it works. This will be a good jumping off point for Wednesday.

padilla410 reviewed (approved):
After @lekoenig added the `Rproj` file back in I was able to run the pipeline and generate the targets that I expected to build (`p1_charNames`)

lekoenig commented on _targets.R:
```diff
@@ -0,0 +1,27 @@
+# here we will define the minor HUCs (hydrologic unit codes) that make up the 
+# DRB to pass to data download functions in dataRetrieval. 
+# The DRB is represented by subregion '0204' (https://water.usgs.gov/GIS/huc_name.html)
+drb_huc8s <- c('02040101','02040102','02040103','02040104','02040105','02040106',
+               '02040201','02040202','02040203','02040204','02040205','02040206','02040207')
```
> How did you make the decision to use HUCs rather than a bounding box? Did this come from the existing DRB scripts?

Good question! My thought was that the bounding box would include some areas outside of the Delaware River Basin, and so would potentially return more data than we'd want if we were only interested in the DRB. I know that `huc` is an attribute included in WQP, so I figured we could use that. _However_, I can certainly imagine cases where the HUC8 field isn't populated in WQP and so wouldn't get returned. Based on this advice from Jordan and Sam, I propose that we use a bounding box instead of the HUC8's here, and perhaps include some documentation that formalizes that reasoning.

jread-usgs replied:
Agreed :)
Another reason to avoid HUC in this case since the example will be pretty generic and re-usable outside of the immediate project needs is that HUC is kind of jargony and hard for lots of users to connect with. My pattern for this type of thing in a past project would be to
1) make a holistic grid of bboxes
2) take my shape that I'm interested in (in your case, the DRB outline, perhaps buffered?)
3) query all bboxes that have _any_ overlap with the shape for sites with data
4) combine all sites from 3 and filter out any sites outside of the shape (point-in-poly analysis)
5) set up a set of calls to make for the data, using some kind of grouping. 

☝️ one reason to create a big set of grids first is that it makes it easier to expand your region of interest without having to re-pull tiles that have data. Not saying this is the best way to do it, but another data point to consider :)

lekoenig replied:
Thanks for this tip, @jread-usgs! I agree that HUC's are jargony and I like that the grid-based approach is potentially more amenable to scaling depending on the use case. I've created an issue to implement the grid-based approach in our pipeline. One thing I plan to keep an eye on, though, is the evenness of the data that gets returned - are the resulting data pulls roughly equal across boxes, or will we want to further partition the site numbers that get returned from our initial grid-based inventory?

jread-usgs replied:
Yes, you are right that the amount of sites (and data) within each grid cell varies a lot. Years ago, the WQP was capable of returning _all sites_ meeting criteria without specifying any spatial filter, but that kind of query now times out so it is necessary to use any one of bbox, state, county, HUC etc. 

My past workflows of this type use a much smaller number of grid cells to make the site query than the number of data calls. For example, I pull temperature sites using a 2° or 5° lat/lon cell, depending on whether I am limiting to all sites (2°) or just lake sites (5°). Since those are pretty large, you'd probably only have a handful of cells in your site query for the DRB. The data query ends up with way more calls because you are then getting the time series data, and I separate those out like others have done and shared with this team by using two thresdholds: number of sites within a single data call can't exceed X and approximate number of total records for all of the sites in the call can't exceed Y, whichever threshold is hit first. I've made this split before by ranking all sites according to `resultCount` and partitioning from there, but a slight variation on that would make things scalable without as much repetition in the calls. If instead you grouped the sites and partitions according to grid cell, then your partitions in the DRB don't get re-scrambled if you add a cell outside of that initial pull (and thus, only the new sites would get pulled).

lindsayplatt commented on 1_fetch/cfg/wqp_codes.yml:
```diff
@@ -0,0 +1,112 @@
+#- "Nitrogen, Kjeldahl"
+#- "Nitrogen, mixed forms (NH3), (NH4), organic, (NO2) and (NO3) as N"
+#- "Nutrient-nitrogen as N"
+#- "Organic Nitrogen as N"
```
Is there a specific reason why some of these are commented out?

lekoenig replied:
They're commented out because I borrowed this yml from `nawqa-wqp` and they were commented out [there](https://github.com/USGS-R/nawqa_wqp/blob/main/1_wqpdata/cfg/wqp_codes.yml) 😄. I haven't checked whether these commented out lines represent valid `characteristicNames` in WQP - @lindsayplatt, do you know? If they're valid names, I'll go ahead and uncomment them.

lindsayplatt replied:
I just tried searching for a couple in [this big list](https://www.waterqualitydata.us/Codes/Characteristicname?mimeType=xml) and don't see them there. Maybe we should just delete?

lekoenig replied:
Thanks for looking into that - I also tried a few basic `dataRetrieval` calls using a couple of the nitrogen ones and was getting errors back. I've opted to delete the commented-out `characteristicNames` from the yml file for now. If we find we're missing any, we can add new `characteristicNames` later.

lindsayplatt commented on 1_fetch.R:
```diff
@@ -0,0 +1,22 @@
+p1_targets_list <- list(
+
+# Get common parameter groups and WQP CharacteristicNames
+tar_target(
```
I think the spacing here is off. All of these things within the `list()` should be indented one.

lindsayplatt commented on 1_fetch.R:
```diff
@@ -0,0 +1,22 @@
+# Get a vector of WQP characteristicNames to match parameter groups of interest
+tar_target(
+  p1_charNames,
+  p1_wqp_params[names(p1_wqp_params) %in% param_groups_select]
```
I think you can simplify this to 

```suggestion
  p1_wqp_params[param_groups_select]
```

lekoenig replied:
Lovely! I've gone ahead and made that change here.

lindsayplatt reviewed (approved):
I was able to successfully run this pipeline and follow the implemented logic. As we discussed in a meeting this morning, the approach to our desired spatial feature and how that maps to our queries will be explored more. I left a few comments but they are about formatting or optimization and are not necessary for this to be considered complete, so I am approving :)

lekoenig commented on _targets.R:
```diff
@@ -0,0 +1,27 @@
+
+# Define the temporal extent of our data pull
+start_date <- '2015-01-01'
+end_date <- '2015-12-31'
```
Thanks, Lindsay. I like this point of thinking about _building_ versus _using_ a targets pipeline. I'll leave `start_date` and `end_date` as is for now, and we can revisit again if needed.

lekoenig commented:
Thanks for your reviews, @padilla410 and @lindsayplatt! I've addressed all of your comments so I'll go ahead and merge this PR since you both approved.
